src/java/org/apache/cassandra/db/compaction/unified/AdaptiveController.java

```diff
@@ -146,93 +147,72 @@ static Controller fromOptions(Environment env,
         int scalingParameter = options.containsKey(STARTING_SCALING_PARAMETER) ? Integer.parseInt(options.get(STARTING_SCALING_PARAMETER)) : DEFAULT_STARTING_SCALING_PARAMETER;
         int[] scalingParameters = new int[32];
         int[] previousScalingParameters = new int[32];
-        long[] flushSizeOverrideMBArray = new long[1];
+        long currentFlushSize = flushSizeOverrideMB << 20;
 
-        //try to read stored options from disk
-        Map<String, Boolean> success = readStoredOptions(scalingParameters, flushSizeOverrideMBArray);
-        if (success.get("scaling_parameters") != null && success.get("scaling_parameters"))
+        File f = getControllerConfigPath();
+        try
         {
-            //readStoredOptions correctly read scaling_parameters so initialize previousScalingParameters as well
-            previousScalingParameters = scalingParameters.clone();
+            JSONParser jsonParser = new JSONParser();
+            JSONObject jsonObject = (JSONObject) jsonParser.parse(new FileReader(f));
+            if (validateStoredScalingParameters(scalingParameters, (JSONArray) jsonObject.get("scaling_parameters")))
```

Comment:
We usually avoid returning data in the parameters if we can; in this case we can construct the array in the method and return null if validation fails.